Create Simplified-payment-verification.md

PhoenixGreen opened the pull request:
## For the easy to read version of this article follow the link.

Article Rendered - https://github.com/PhoenixGreen/Cypherpunktimes-articles/blob/PhoenixGreen-draft-1/moving-forward-series/simple-payment-verification.md

## Social media clips:
Raw video files - https://drive.google.com/drive/folders/1yx0h5B1rdFO3lXy6gC9eoTIqwwqyfNzz

Clip 1 - SPV challenges centralised lite client conventions 
Clip 2 - Non-custodial mobile wallet
Clip 3 - Decred SPV how to be fast private and secure
Clip 4 - compact block filters
Clip 5 - Block Header Commitments consensus upgrade 2020
Clip 6 - SPV vs Lite clients
Clip 7 - Lite client centralised setup
Clip 8 - Decred interchangeable wallets
Clip 9 - Decred SPV a shining example of decentralisation

## Social media posts:

* Main video post - https://x.com/decredproject/status/1748450782049894692
* Clip 1 post - https://x.com/DecredSociety/status/1748094644083118383
* Clip 2 post - https://x.com/DecredSociety/status/1748450803184980009
*

PhoenixGreen commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,31 @@
+# What is Simple Payment Verification (SPV)  - Moving forward (better title needed)
```
What should I change the title to?

xaur replied:
`Can a Crypto Wallet be Light AND Secure? SPV Done Right - Moving Forward`

Just a quick one, better ideas welcome

matthawkins90 replied:
`Can a Crypto Wallet be Lightweight, Private AND Secure? The SPV Model Leads the Way"

PhoenixGreen replied:
When I've finished with a comment, do I press the "Resolve conversation" button or is there something else I should do first?

matthawkins90 replied:
"Resolve conversation" is fine! Anyone can unresolve it if they really wanted to 😂

xaur replied:
When you "resolve" it collapses the discussion to take less space and mark it as "done" or "processed". All comments are still accessible if you expand, and it can be unresolved if needed. Just an attention management tool.

xaur commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,31 @@
+There is a need for mobile wallets that are fast to set up, easy to use and very secure. All too often, mobile wallets sacrifice the security of a fully validating wallet for easy of use. SPV challenges this convention by providing a high level of security whilst improving the user experience. 
+
+## Fast and Small
+If you’re looking for a non-custodial mobile wallet, Simple Payment Verification (SPV) is an idea solution.  SPV is designed to be intentionally lightweight, as it interacts with the blockchain by only fetching and downloads the data, that is absolutely necessary. This includes block headers, block filters, and the full blocks that contain transaction data, directly associated with the wallet. As ...
```
Let's have a blank line after each heading. It should look the same when rendered but it reads better in text editors and monospace fonts.

Personally I use *two* blank lines before a heading and one after to get even better UX in monospace text editors, but it's up to you.

xaur commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,31 @@
+## Fast and Small
+If you’re looking for a non-custodial mobile wallet, Simple Payment Verification (SPV) is an idea solution.  SPV is designed to be intentionally lightweight, as it interacts with the blockchain by only fetching and downloads the data, that is absolutely necessary. This includes block headers, block filters, and the full blocks that contain transaction data, directly associated with the wallet. As ...
+
+## SPV Filters
```
I haven't seen the use of term `SPV Filters` in the wild. They are referred to as "compact block filters" or "compact filters" or "block filters" or just "cfilters" in dev talk.

https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,31 @@
+## Benefits and drawbacks
+To conclude, SPV and Lite clients manage risk and functionality in various ways. If you require the highest assurances, there is still no better option than running a fully validating wallet. But if you’re seeking a lightweight alternative to interact with the blockchain, SPV is a pretty solid option that has seen many improvements over the last few years to increase security and privacy. 
+
+As it currently stands, the main drawbacks with using lite clients and custodial wallets over a properly implemented SPV system. Revolve around the ability to validate the data the wallet receives and the need to trust third-party services or servers to process the information required. 
```
These two sentences are not full sentences, and need to be combined into one sentence. However, when you do that, it becomes a very long run-on sentence. Maybe just cut out the fluff and shorten it to:

* "As it currently stands, the main drawbacks with using lite clients revolve around having to trust centralized third-party services instead of the decentralized network."

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,31 @@
+
+As it currently stands, the main drawbacks with using lite clients and custodial wallets over a properly implemented SPV system. Revolve around the ability to validate the data the wallet receives and the need to trust third-party services or servers to process the information required. 
+
+There is a need for lightweight wallets and currently Decred’s SPV is a shining example of how this can be achieved whilst still being in full control of your coins, data, and privacy.
```
change to:
* "while still allowing users to be in full control of their coins, data, and privacy."
* or maybe "while still allowing you to be in full control of your coins, data, and privacy."

matthawkins90 reviewed (commented):
On the whole, I think this is organized very well. I made in-line comments, but here are two general thoughts:

* The discussion of Attack Vectors needs to be more than a sidenote. Read more in my in-line comments, but it feels incomplete as is. I think the organization seems off, and maybe it deserves its own section.
* Make sure you're referencing the original DCP-5 very carefully. Don't deviate or contradict things from the DCP.
    * It's called `Simplified Payment Verification`, not `Simple Payment Verification`
    * When you're talking about the security model, it's called `The SPV model`.
    * When you're talking about a wallet, it's called `An SPV wallet`.
        * Nodes validate and serve blockchain data. A wallet just downloads blocks. So an SPV wallet makes sense, but an SPV node does not.

xaur commented:
> I think "Block Filters" helps non-technical readers the most.

Agreed. This tech is brilliant but everything sounds too cryptic. To help people understand and appreciate it we need to dumb it down.

If it may help, here's a quick simplified explainer.

To make these "filters" less alien, we can ask why they are called so and how they work. Imagine we are a light wallet. Our job is to track user's funds, more specifically to track addresses and transactions (having those we can compute the balances). Where do we get the data from? The blocks. We need data from the blocks. But we have restrictions. Since we want to be light and fast, we don't want to download and store the entire blockchain. We also want to be private/secure, so we can't ask centralized servers "give me data for this address" and help them map our stuff.

So we need some smart way to get data from the blocks to meet those restrictions. Here's what we do. First step is to identify *which blocks* have our data. We connect to full nodes directly and download these tiny smart pieces of data called "filters", one filter for each block. Why "filters", what are we "filtering"? We are filtering *blocks*. For each filter, we take our wallet data and throw it at the filter to see if its associated full block has any data we're interested in. If the answer is no, we skip that block (i.e. never download or process it), or we "filter it out". If the answer is yes, the block associated with the filter has something relevant for our wallet, with very high probability. So we move to the second step - fetch that full block, find our data in it, possibly discover new addresses/transactions we need to track, and update our records. Then move to the next filter and repeat the process.

In a way, we "filter through" a very lightweight reflection of the full chain, and fetch our data as we discover where it is.

If "filter" still sounds alien, another word I came up with is "matcher", i.e. we test wallet data against compact matchers and if they say yes, we fetch the full block. This is just for internal understanding, don't use "matcher" in writing.

Filters for the entire ~12 GB chain take about 200 MB. And since we store them locally, we can re-run through them later to recover from bugs (e.g. we suspect the balance is off).

This short page introduces it pretty well too: https://bitcoinops.org/en/topics/compact-block-filters/

PhoenixGreen commented:
Hi all, I've made the above changes and also added a "simple example" section based on xaur example. I've removed the attack vector elements and have tried to be consistent with the technical conventions of the DCP doc. I've got a little more editing to do today but generally I'm happy with the flow of this document. If you do notice any glaring issues, please let me know. I'm aiming to do the recording for this piece tomorrow, if there ain't any show stoppers. Thanks again for all your feedback.

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+## Can a Crypto Wallet be Lightweight, Private, AND Secure? The SPV Model Leads the Way.
+
+People need mobile wallets that are fast to set up, easy to use and very secure. All too often, mobile wallets make extreme security sacrifices in order to slightly improve ease of use. Including, the provision for a centralised node to serve requests to the wallet software. This is a trust model that processes all wallet information, sent and received, through a single point of failure. SPV chal ...
```
* This is not a complete sentence: "Including, the provision for a centralized node..." 
    * Change the sentences to this:
        * One example of a bad security practice is when a wallet has to trust a centralised node.
* This first paragraph is the very first time you mention SPV (other than the title, which is perfectly fine to leave as-is.) Therefore, you should spell out SPV the first time you use it in your article.
    * Change to:
        * A wallet using a Simplified Payment Verification (SPV) model challenges this convention by providing...

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+## Fast and Small
+
+If you’re looking for a non-custodial mobile wallet, Simplified Payment Verification (SPV) is an ideal solution. SPV is designed to be intentionally lightweight, as it interacts with the blockchain by only fetching and downloading the data, that is absolutely necessary. This includes block headers, block filters, and the full blocks that contain transaction information directly associated with th ...
```
* Since we moved the full spelling to the first paragraph, we can remove it here. But I'd clarify it a bit.
    * Change to
        * If you're looking for a non-custodial mobile wallet, the ideal solution is one that uses the SPV model. SPV wallets are designed...

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+If you’re looking for a non-custodial mobile wallet, Simplified Payment Verification (SPV) is an ideal solution. SPV is designed to be intentionally lightweight, as it interacts with the blockchain by only fetching and downloading the data, that is absolutely necessary. This includes block headers, block filters, and the full blocks that contain transaction information directly associated with th ...
+
+## How does Decred’s SPV work (a simple example)
```
How does Decred's SPV wallet work? (a simple example)

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+## How does Decred’s SPV work (a simple example)
+
+Imagine we are a light wallet. Our job is to track user's funds, more specifically to track addresses and transactions (having those we can compute the balances). Where do we get the data from? The blocks. We need data from the blocks. But we have restrictions. Since we want to be light and fast, we would rather not download and store the entire blockchain. We also would like to be private and se ...
```
* Use an em dash (—), or simply just two dashes (--) instead of a comma here:
    *  Our job is to track user's funds—more specifically...

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+## Compact Block Filters
+
+SPV relies on the ability to identify and retrieve relevant transactions from the blockchain. Filters act like a search mechanism to help an SPV wallet find transactions it should care about. Decred’s Compact filters allow SPV nodes to retrieve this information without having to reveal or leak information regarding coin ownership. For instance, SPV nodes can ask full node peers for blocks that co ...
```
* The SPV model relies on the ability...
* Decred's compact filters allow SPV wallets to retrieve this information...
* For instance, SPV wallets can ask full node peers...


"SPV nodes" do not exist.

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+## Benefits and drawbacks
+
+To conclude, SPV and Lite clients manage risk and functionality in various ways. If you require the highest assurances, there is still no better option than running a fully validating wallet. But if you’re seeking a lightweight alternative to interact with the blockchain, SPV is a pretty solid option that has seen many improvements over the last few years to increase security and privacy. 
```
* It's okay to say "SPV and Lite clients", because the word "SPV" is modifying "client". So feel free to live this alone.
* But you should change:
    * ...to interact with the blockchain, an SPV wallet is a pretty solid option...

matthawkins90 commented on moving-forward-series/simple-payment-verification.md:
```diff
@@ -0,0 +1,50 @@
+
+As it currently stands, the main drawbacks with using lite clients revolve around having to trust centralised third-party services instead of the decentralised network.
+
+There is a need for lightweight wallets and currently Decred’s SPV is a shining example of how this can be achieved while still allowing users to be in full control of their coins, data, and privacy.
```
* Decred's SPV model is a shining example...

matthawkins90 reviewed (commented):
Great changes. Reads very well now. I only had a handful of nitpicky notes.

While it might be perfectly accurate to just say "SPV" by itself, I think it just _feels_ better to use "SPV" followed by a noun. So I suggested that where I felt like it would read better.

PhoenixGreen commented:
Perfect! Thank you for taking the time to do this final review, much appreciated. All changes have been made, time to record 👍

xaur commented:
Moved in line with the new naming scheme:

https://github.com/PhoenixGreen/Cypherpunktimes-articles/blob/main/articles/20240118.what-is-spv.md
